File: cpp/lld/board-games/tictactoe/TicTacToeEngine.cpp
#include <vector>
#include <string>
#include <optional>
#include "game_status.h"
#include "position.h"
#include "piece.h"
#include "player.h"
#include "move.h"

class TicTacToeEngine {
    std::vector<std::vector<std::optional<Piece>>> grid;
    std::vector<Player> players = {
        Player{"Player 1", Piece{"X", "#818CF8"}},
        Player{"Player 2", Piece{"O", "#F59E0B"}},
    };
    int size;
    int currentPlayerIndex;
    GameStatus status;
    std::vector<Move> moveHistory;
    std::optional<Player> winner;

public:
    TicTacToeEngine(int s = 3)
        : size(s),
          currentPlayerIndex(0),
          status(GameStatus::NOT_STARTED) {
        start();
    }

    void start() {
        status = GameStatus::IN_PROGRESS;
        currentPlayerIndex = 0;
        moveHistory.clear();
        winner = std::nullopt;
        grid.assign(
            size,
            std::vector<std::optional<Piece>>(
                size, std::nullopt
            )
        );
    }

    auto getGrid() const { return grid; }

    Player getCurrentPlayer() const {
        return players[currentPlayerIndex];
    }

    GameStatus getStatus() const {
        return status;
    }

    std::optional<Player> getWinner() const {
        return winner;
    }

    bool makeMove(
        const Player& player,
        const Position& to
    ) {
        if (status != GameStatus::IN_PROGRESS) {
            return false;
        }
        if (to.row < 0 || to.row >= size ||
            to.col < 0 || to.col >= size) {
            return false;
        }
        if (grid[to.row][to.col].has_value()) {
            return false;
        }

        grid[to.row][to.col] = player.piece;
        moveHistory.push_back({player, to});

        if (checkWin(to)) {
            status = GameStatus::COMPLETED;
            winner = player;
            return true;
        }

        bool full = true;
        for (auto& row : grid) {
            for (auto& cell : row) {
                if (!cell.has_value()) {
                    full = false;
                }
            }
        }
        if (full) {
            status = GameStatus::DRAW;
            return true;
        }

        currentPlayerIndex =
            (currentPlayerIndex + 1) % players.size();
        return true;
    }

private:
    bool checkWin(const Position& pos) {
        auto piece = grid[pos.row][pos.col];
        if (!piece.has_value()) return false;

        int dirs[4][2] = {
            {0, 1}, {1, 0}, {1, 1}, {1, -1}
        };
        for (auto& d : dirs) {
            int count = 1;
            for (int dir : {-1, 1}) {
                for (int i = 1; i < size; i++) {
                    int r = pos.row + d[0] * i * dir;
                    int c = pos.col + d[1] * i * dir;
                    if (r < 0 || r >= size ||
                        c < 0 || c >= size) break;
                    auto cell = grid[r][c];
                    if (cell.has_value() &&
                        cell->symbol == piece->symbol) {
                        count++;
                    } else {
                        break;
                    }
                }
            }
            if (count >= size) return true;
        }
        return false;
    }
};

File: typescript/lld/board-games/tictactoe/TicTacToeEngine.ts
import { GameStatus } from './GameStatus';
import { Position } from './Position';
import { Piece } from './Piece';
import { Player } from './Player';
import { Move } from './Move';

export class TicTacToeEngine {
    private grid: (Piece | null)[][];
    private currentPlayerIndex = 0;
    private status = GameStatus.NOT_STARTED;
    private moveHistory: Move[] = [];
    private winner: Player | null = null;

    readonly players: Player[] = [
        { name: 'Player 1', piece: { symbol: 'X', color: '#818CF8' } },
        { name: 'Player 2', piece: { symbol: 'O', color: '#F59E0B' } },
    ];
    readonly size: number;

    constructor(size: number = 3) {
        this.size = size;
        this.grid = Array.from({ length: size }, () =>
            Array.from({ length: size }, () => null)
        );
    }

    start(): void {
        this.status = GameStatus.IN_PROGRESS;
        this.currentPlayerIndex = 0;
        this.moveHistory = [];
        this.winner = null;
        this.grid = Array.from({ length: this.size }, () =>
            Array.from({ length: this.size }, () => null)
        );
    }

    getGrid(): (Piece | null)[][] {
        return this.grid.map(row => [...row]);
    }

    getCurrentPlayer(): Player {
        return this.players[this.currentPlayerIndex];
    }

    getStatus(): GameStatus {
        return this.status;
    }

    getWinner(): Player | null {
        return this.winner;
    }

    makeMove(player: Player, to: Position): boolean {
        if (this.status !== GameStatus.IN_PROGRESS) return false;
        if (to.row < 0 || to.row >= this.size ||
            to.col < 0 || to.col >= this.size) return false;
        if (this.grid[to.row][to.col] !== null) return false;
        if (player !== this.players[this.currentPlayerIndex]) return false;

        this.grid[to.row][to.col] = player.piece;
        this.moveHistory.push({ player, to });

        if (this.checkWin(to)) {
            this.status = GameStatus.COMPLETED;
            this.winner = player;
            return true;
        }

        if (this.grid.every(row => row.every(c => c !== null))) {
            this.status = GameStatus.DRAW;
            return true;
        }

        this.currentPlayerIndex =
            (this.currentPlayerIndex + 1) % this.players.length;
        return true;
    }

    private checkWin(pos: Position): boolean {
        const piece = this.grid[pos.row][pos.col];
        if (!piece) return false;

        const dirs = [[0, 1], [1, 0], [1, 1], [1, -1]];
        for (const [dr, dc] of dirs) {
            let count = 1;
            for (const dir of [-1, 1]) {
                for (let i = 1; i < this.size; i++) {
                    const r = pos.row + dr * i * dir;
                    const c = pos.col + dc * i * dir;
                    if (r < 0 || r >= this.size ||
                        c < 0 || c >= this.size) break;
                    const cell = this.grid[r][c];
                    if (cell?.symbol === piece.symbol) count++;
                    else break;
                }
            }
            if (count >= this.size) return true;
        }
        return false;
    }
}